src/helper/tools/text-tool.js

```diff
@@ -0,0 +1,326 @@
+        this.onKeyUp = this.handleKeyUp;
+        this.onKeyDown = this.handleKeyDown;
+
+        this.textBox = null;
+        this.guide = null;
+        this.colorState = null;
+        this.mode = null;
+        this.active = false;
+        this.lastTypeEvent = null;
+
+        // If text selected and then activate this tool, switch to text edit mode for that text
+        // If double click on text while in select mode, does mode change to text mode? Text fully selected by default
+    }
+    getBoundingBoxHitOptions () {
+        return {
+            segments: true,
+            stroke: true,
+            curves: true,
+            fill: true,
+            guide: false,
+            match: hitResult =>
+                (hitResult.item.data && hitResult.item.data.isHelperItem) ||
+                hitResult.item.selected, // Allow hits on bounding box and selected only
+            tolerance: TextTool.TOLERANCE / paper.view.zoom
+        };
+    }
+    getTextEditHitOptions () {
+        return {
+            class: paper.PointText,
+            segments: true,
+            stroke: true,
+            curves: true,
+            fill: true,
+            guide: false,
+            match: hitResult => hitResult.item && !hitResult.item.selected, // Unselected only
+            tolerance: TextTool.TOLERANCE / paper.view.zoom
+        };
+    }
+    /**
+     * Called when the selection changes to update the bounds of the bounding box.
+     * @param {Array<paper.Item>} selectedItems Array of selected items.
+     */
+    onSelectionChanged (selectedItems) {
+        this.boundingBoxTool.onSelectionChanged(selectedItems);
+    }
+    // Allow other tools to cancel text edit mode
+    onTextEditCancelled () {
+        this.endTextEdit();
+        if (this.textBox) {
+            this.mode = TextTool.SELECT_MODE;
+            this.textBox.selected = true;
+            this.setSelectedItems();
+        }
+    }
+    /**
+     * Called when the view matrix changes
+     * @param {paper.Matrix} viewMtx applied to paper.view
+     */
+    onViewBoundsChanged (viewMtx) {
+        if (this.mode !== TextTool.TEXT_EDIT_MODE) {
+            return;
+        }
+        const matrix = this.textBox.matrix;
+        this.element.style.transform =
+            `translate(0px, ${this.textBox.internalBounds.y}px)
+            matrix(${viewMtx.a}, ${viewMtx.b}, ${viewMtx.c}, ${viewMtx.d},
+            ${viewMtx.tx}, ${viewMtx.ty})
+            matrix(${matrix.a}, ${matrix.b}, ${matrix.c}, ${matrix.d},
+            ${matrix.tx}, ${matrix.ty})`;
+    }
+    setColorState (colorState) {
+        this.colorState = colorState;
+    }
+    handleMouseMove (event) {
+        const hitResults = paper.project.hitTestAll(event.point, this.getTextEditHitOptions());
+        if (hitResults.length) {
+            document.body.style.cursor = 'text';
+        } else {
+            document.body.style.cursor = 'auto';
+        }
+    }
+    handleMouseDown (event) {
```

Comment:
There is some seriously complicated logic in this function. I'd like for us to come back to this and break it into more clear state transitions. It took me ~20 minutes to understand the entire FROM => TO transitions!